### PR TITLE
Restructure licensing profile

### PR DESCRIPTION
**Pull request by armintaenzertng**
This PR summarizes the discussion we had on the tech team call yesterday (June 27).
This moves all classes used for more in-depth modeling of licenses into the `ExpandedLicensing` profile, keeping only `AnyLicenseInfo`, `LicenseExpression` as well as a new `SimpleCustomLicense` class in the `LicensingProfile`.

The new structure is open to discussion. I believe some descriptions have to be rewritten.

### CODE REVIEW

**Comment by zvr** [comment]
I propose `SimpleCustomLicense` be renamed to `SimpleLicensingText` or something.

Then this can be used for License texts as well as for Addition texts. Otherwise, we would need a `SimpleCustomAddition` in the Licensing profile, which will be identical to `SimpleLicense` (only a `string` property for the text).

File: model/Licensing/Classes/SimpleCustomLicense.md
Comment (zvr) [inline model/Licensing/Classes/SimpleCustomLicense.md:3]: `SimpleLicensingText`

File: model/Licensing/Classes/SimpleCustomLicense.md
Comment (zvr) [inline model/Licensing/Classes/SimpleCustomLicense.md:11]: same changes

File: model/Licensing/Classes/SimpleCustomLicense.md
Comment (zvr) [inline model/Licensing/Classes/SimpleCustomLicense.md:17]: `SimpleLicensingText`

**Review by zvr — CHANGES_REQUESTED** [review]

File: model/Licensing/Classes/SimpleCustomLicense.md
Comment (armintaenzertng) [reply model/Licensing/Classes/SimpleCustomLicense.md]: I took the liberty to also change the corresponding property name, I hope that was in your interest, too.

**Review by zvr — APPROVED** [review]

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (goneall) [inline model/Licensing/Classes/SimpleLicensingText.md:18]: I'm not sure if we want `SimpleLicensingText` to be a subclass of `AnyLicenseInfo` - if someone is making a declared license or concluded license, should we restrict it to always be either a license expression string or an expanded license?  If you want to reference a custom license or addition, I think you would want to use a license expression string with the ID of the `SimpleLicensingText` and keep the `SimpleLicensingText` as a direct subclass of `Element`.  Otherwise, it would be more complex to parse in the simple case and it would be confusing when to use the `SimpleLicensingText` or the `CustomLicense`.

File: model/Licensing/Licensing.md
Comment (goneall) [inline model/Licensing/Licensing.md:15]: This makes me think we should rename `ExpandedLicense` to `Licensing` with the same description above and rename the current `Licensing` to `SimpleLicensing`.

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (goneall) [inline model/Licensing/Classes/SimpleLicensingText.md:23]: Could we just use the `licenseText` property and move it back from the `ExpandedLicense` to `Licensing`?  It seems like it has similar semantics

**Review by goneall — COMMENTED** [review]
More of a request for a discussion on my comments.

**Comment by goneall** [comment]
Thanks @armintaenzertng for creating this - overall structure looks good based on our discussion.  A couple of review comments to consider / discuss.

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (armintaenzertng) [reply model/Licensing/Classes/SimpleLicensingText.md:18]: I agree that `SimpleLicensingText` is better off subclassing `Element`.

Comment (zvr) [reply model/Licensing/Classes/SimpleLicensingText.md:18]: I also agree.

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (armintaenzertng) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: `licenseText` is used in the ExpandedLicense profile, so if we move it we would have to reference it from another profile. That would be a first, I think. I'm not sure if we currently support this. @zvr?
I intentionally renamed this so that we don't get the same property name in two different profiles (we had some discussion a while back that we would try to avoid that if possible).

File: model/Licensing/Licensing.md
Comment (goneall) [reply model/Licensing/Licensing.md:15]: > Maybe it will be clearer to have SimpleLicensing and ExpandedLicensing profiles.

I like this naming.  In `SimpleLicensing` we could refer to `ExpandedLicensing` for the detailed definitions.

I would suggest waiting on any renaming until @swinslow has a chance to weigh in.

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (goneall) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: > so if we move it we would have to reference it from another profile

I think this is OK, but I'll defer to @zvr for a definitive answer on this

> I intentionally renamed this so that we don't get the same property name in two different profiles (we had some discussion a while back that we would try to avoid that if possible).

Agree we should not use the same property name in 2 different profiles - I would propose referring to the property in a different profile as long as it is supported

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (armintaenzertng) [reply model/Licensing/Classes/SimpleLicensingText.md]: done

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (zvr) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: We could do it (as we could do almost anything), especially since there is no way to use `ExpandedLicensing` profile without using the `SimpleLicensing` one.
But I think it would introduce complexity in tooling (and in spec-parser) to have classes in a profile having properties from other profiles. Remember, they all end up with IRIs with namespaces.

To Gary's last point "we should not use the same property name in 2 different profiles", we're already past this stage, as you all know: cf. `locator`.

Comment (goneall) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: > But I think it would introduce complexity in tooling (and in spec-parser) to have classes in a profile having properties from other profiles. Remember, they all end up with IRIs with namespaces.

@zvr We already have `Software` classes referencing other classes (e.g. `SoftwareArtifact.concludedLicense` refers to a property in a namespace other than `SoftwareArtifact`).  It this situation more complex?

Since the `licenseText` and `simpleLicensingText` are semantically the same, it would be useful for the RDF users to use the same property name.  BTW - this is solveable with different property names by using `owl:sameAs`, but it would not be as straight-forward as just using the same property name.

Comment (zvr) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: @goneall the example you cited will be replaced by a Relationship :-P

Do we have another?

Comment (goneall) [reply model/Licensing/Classes/SimpleLicensingText.md:23]: `/Core/PositiveIntegerRange` referenced from `Software/Snippet` is another example.

**Comment by goneall** [comment]
@armintaenzertng - Pinging you on this one - can you go back through the review comments and update what you agree with and (re)comment on any outstanding?

I just bumped into this in my java code generator and would like to merge.

File: model/Licensing/Licensing.md
Comment (armintaenzertng) [reply model/Licensing/Licensing.md:15]: I implemented the renaming.

File: model/Licensing/Classes/SimpleLicensingText.md
Comment (armintaenzertng) [reply model/Licensing/Classes/SimpleLicensingText.md]: The `licenseText` property is now moved to `SimpleLicensing`, while also still used in the `ExpandedLicensing` profile.

File: model/ExpandedLicense/Classes/WithAdditionOperator.md
Comment (armintaenzertng) [reply model/ExpandedLicense/Classes/WithAdditionOperator.md]: fixed

**Comment by armintaenzertng** [comment]
@goneall, @zvr, I implemented all outstanding change requests, please take another look.

**Review by goneall — APPROVED** [review]
Thanks @armintaenzertng

**Comment by armintaenzertng** [comment]
Found and fixed another little bug.

**Comment by swinslow** [comment]
I have a few concerns about the way this is currently addressed. I'm not going to stand in the way of merging it as a next draft, but from an initial read, these are the points that I think should be addressed prior to the actual 3.0 release:

* As currently structured, it looks like for validation purposes the only requirement is that `LicenseExpression` should be an xsd:string. My assumption is that it additionally would only be valid if that string conforms to the [License Expression Syntax](https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/). I'm not sure how we express that in the spdx-3-model repo syntax, but I think that's an important thing for us to retain if all the pieces of the model for license combinations are being moved to a separate profile.

* Relatedly, I think part of the expression being valid is to be able to ensure that the license identifiers are valid, e.g. that any non-LicenseRef's are actual identifiers on the License List. I'm not sure whether that's feasible to do under the current structure without leveraging the ExpandedLicensing profile, if ListedLicense and ListedLicenseException are over in that profile.

* I am worried that it's confusing to have a `simpleLicensingText` class in the SimpleLicensing profile which is comparable to, but distinct from, `CustomLicense` in the ExpandedLicensing profile.

**Comment by swinslow** [comment]
On the first point above, I should note that @goneall has filed #440 to track clarifying this.

**Comment by goneall** [comment]
Per discussion with @swinslow - I'll go ahead and merge this PR and we can discuss the outstanding issues raised [in the comment above](https://github.com/spdx/spdx-3-model/pull/399#issuecomment-1654366122) after merging.

**Pull request merged**
